node.go

```diff
@@ -293,6 +435,8 @@ func (node *Node) validate() error {
 	if node == nil {
 		return errors.New("node cannot be nil")
 	}
+	node.mtx.RLock()
```

Comment:
This read lock will lock the whole validate() function, however comparing with other functions, it's different. Other functions we only change to the thread safe basic CRUD function, but we don't lock the whole function. 

My question is why are we doing this and how do we determine when to lock the whole function, when not to. For other function which we don't lock the whole function, it's still not thread safe right?